File: lab23.cs
using System;
using System.Collections.Generic;
using System.Linq;

public interface IDigitProcessor
{
    double GetDigitPercentage();
    int GetTotalCharacterCount();
}

// Клас-рядок
public class StringElement
{
    private string data;

    public StringElement(string str)
    {
        data = str;
    }

    public string GetString() => data;

    public int Length() => data.Length;

    public bool HasDigits() => data.Any(char.IsDigit);

    public int CountDigits() => data.Count(char.IsDigit);
}

// Клас-контейнер
public class TextContainer : IDigitProcessor
{
    private List<StringElement> textElements = new List<StringElement>();

    public void AddString(string str) => textElements.Add(new StringElement(str));

    public void RemoveString(string str)
    {
        textElements.RemoveAll(elem => elem.GetString() == str);
    }

    public void ClearText() => textElements.Clear();

    public string GetLargestString()
    {
        return textElements.Count == 0 ? "" : textElements.OrderByDescending(e => e.Length()).First().GetString();
    }

    public double GetDigitPercentage()
    {
        int totalChars = GetTotalCharacterCount();
        if (totalChars == 0) return 0.0;

        int digitCount = textElements.Sum(elem => elem.CountDigits());

        return (double)digitCount / totalChars * 100.0;
    }

    public int GetTotalCharacterCount() => textElements.Sum(elem => elem.Length());

    public void PrintText()
    {
        foreach (var elem in textElements)
        {
            Console.WriteLine(elem.GetString());
        }
    }
}

// Тестування
public class Program
{
    public static void Main()
    {
        TextContainer text = new TextContainer();
        text.AddString("Hello123");
        text.AddString("Programming");
        text.AddString("C# is fun!");
        text.AddString("2025");

        Console.WriteLine("Збережений текст:");
        text.PrintText();
        
        Console.WriteLine($"\nНайбільший рядок: {text.GetLargestString()}");
        Console.WriteLine($"Відсоток цифр у тексті: {text.GetDigitPercentage()}%");
        Console.WriteLine($"Загальна кількість символів: {text.GetTotalCharacterCount()}");

        text.RemoveString("2025");
        Console.WriteLine("\nПісля видалення '2025':");
        text.PrintText();

        text.ClearText();
        Console.WriteLine("\nПісля очищення тексту:");
        text.PrintText();
    }
}

File: lab23.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cctype>

using namespace std;

// Інтерфейс для роботи з цифрами
class IDigitProcessor {
public:
    virtual double getDigitPercentage() const = 0;
    virtual size_t getTotalCharacterCount() const = 0;
    virtual ~IDigitProcessor() = default;
};

// Клас-рядок
class StringElement {
private:
    string data;
public:
    StringElement(const string& str) : data(str) {}

    string getString() const {
        return data;
    }

    size_t length() const {
        return data.size();
    }

    bool hasDigits() const {
        return any_of(data.begin(), data.end(), ::isdigit);
    }

    size_t countDigits() const {
        return count_if(data.begin(), data.end(), ::isdigit);
    }
};

// Клас-контейнер
class TextContainer : public IDigitProcessor {
private:
    vector<StringElement> textElements;
public:
    void addString(const string& str) {
        textElements.emplace_back(str);
    }

    void removeString(const string& str) {
        textElements.erase(
            remove_if(textElements.begin(), textElements.end(),
                      [&str](const StringElement& elem) { return elem.getString() == str; }),
            textElements.end());
    }

    void clearText() {
        textElements.clear();
    }

    string getLargestString() const {
        if (textElements.empty()) return "";
        return max_element(textElements.begin(), textElements.end(),
                           [](const StringElement& a, const StringElement& b) {
                               return a.length() < b.length();
                           })->getString();
    }

    double getDigitPercentage() const override {
        size_t totalChars = getTotalCharacterCount();
        if (totalChars == 0) return 0.0;

        size_t digitCount = 0;
        for (const auto& elem : textElements) {
            digitCount += elem.countDigits();
        }

        return (static_cast<double>(digitCount) / totalChars) * 100.0;
    }

    size_t getTotalCharacterCount() const override {
        size_t totalChars = 0;
        for (const auto& elem : textElements) {
            totalChars += elem.length();
        }
        return totalChars;
    }

    void printText() const {
        for (const auto& elem : textElements) {
            cout << elem.getString() << endl;
        }
    }
};

// Тестування
int main() {
    TextContainer text;
    text.addString("Hello123");
    text.addString("Programming");
    text.addString("C++ is fun!");
    text.addString("2025");

    cout << "Збережений текст:" << endl;
    text.printText();
    
    cout << "\nНайбільший рядок: " << text.getLargestString() << endl;
    cout << "Відсоток цифр у тексті: " << text.getDigitPercentage() << "%" << endl;
    cout << "Загальна кількість символів: " << text.getTotalCharacterCount() << endl;

    text.removeString("2025");
    cout << "\nПісля видалення '2025':" << endl;
    text.printText();

    text.clearText();
    cout << "\nПісля очищення тексту:" << endl;
    text.printText();

    return 0;
}
